certbot-apache/certbot_apache/_internal/prefetch_ocsp.py

```diff
@@ -143,9 +219,9 @@ def _ocsp_ttl(self, next_update):
         :rtype: int
         """
 
-        if next_update:
-            now = time.time()
-            res_ttl = int(time.mktime(next_update.timetuple()) - now)
+        if next_update is not None:
+            now = datetime.fromtimestamp(time.time())
```

Comment:
My brain is a bit scrambled after reading about dates and times in Python so I'd double check everything I'm saying here, but I _think_ we want to use `utcfromtimestamp` here.

The datetime objects being used both in the variable `now` and in `next_update` regardless of the OCSP backend used are [naive rather than aware](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects). Maybe we should fix this, but I _think_ always using naive objects is fine as long as we're sure all of the datetime objects are in the same timezone.

I also _think_ [datetime.fromtimestamp](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp) converts the timestamp into local time while [datetime.utcfromtimestamp](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp) keeps it as UTC.

I checked this by setting a breakpoint and comparing the calculation done here to one done on a site like https://www.timeanddate.com/ and using `utfcfromtimestamp` at least matches the total number of seconds there.

Also, do you know what Apache's behavior is if we get this wrong and the time in the next update field passes before the TTL we give Apache? Depending on that and how confident we are this code is right, we may want to give ourselves more than an hour of leeway to prevent Apache from serving bad OCSP responses in case something goes wrong.

Reply:
> My brain is a bit scrambled after reading about dates and times in Python so I'd double check everything I'm saying here, but I think we want to use utcfromtimestamp here.

You are right, though I'm going with a third option (because of course) here: `datetime.utcnow()` which is essentially the same is `datetime.utcfromtimestamp(time.time())`.

> Also, do you know what Apache's behavior is if we get this wrong and the time in the next update field passes before the TTL we give Apache? 

Yeah, It will happily serve the expired OCSP staple. There's no validation in between. So you have the right call here, we should be careful with the timestamps as we're setting the internal TTL so close to the expiry.